### PR TITLE
new algorithm for scheduler HA predicate

### LINKED ISSUE #235: A new algorithm for tidb scheduler HA predicate
Now the algorithm of [HA predicate](https://github.com/pingcap/tidb-operator/blob/master/pkg/scheduler/predicates/ha.go#L54) is complicated and hard to understand:

1. First, we sort all the nodes we get from kube-scheduler by how many same kind of pod it contains, find the nodes that have least pods.
2. When scheduling the first replicas pods, we must ensure no previous pods on the nodes.
3. For later pods, we choose the nodes that have least pods.

So we need an easier solution. If a user cannot access the system, it is – from the users' point of view – unavailable. In TiDB Operator, we say HA, it means the whole system is still available when there is **one** Kube node down. 

Discussed with @tennix @onlymellb and @xiaojingchen. The new algorithm is:

- ~~multiple pods are not allowed to be scheduled concurrently in TiDB Operator: lower ordinal pods must be scheduled before the current pod~~
- return the node to kube-scheduler if there is only one node and the pod's pvc is bound 
- return these nodes that have least pods and its pods count is less than replicas/2 to kube-scheduler
- let kube-scheduler to make the final decision

Then all the pods count of Kube nodes is equal or less then replicas / 2, it is available(the raft group is still in quorum.) when this node is down.

### PR DESCRIPTION
**Pull request by weekface**
fixes #235 

@tennix @onlymellb @xiaojingchen PTAL

### CODE REVIEW

File: pkg/scheduler/predicates/ha_test.go [hunk 1]
```diff
@@ -39,229 +40,327 @@ func TestMapNil(t *testing.T) {
...
 			expectFn: func(nodes []apiv1.Node, err error) {
-				g.Expect(strings.Contains(err.Error(), "strconv.Atoi: parsing")).To(Equal(true))
+				g.Expect(err).To(HaveOccurred())
+				g.Expect(strings.Contains(err.Error(), "get tidbcluster failed")).To(BeTrue())
 			},
 		},
 		{
-			name:      "one pod, podName is wrong",
-			ordinal:   0,
+			name:      "one node, return zero node",
```
Comment (xiaojingchen) [inline pkg/scheduler/predicates/ha_test.go]:
```suggestion
			name:      "zero node, return zero node",
```

File: pkg/scheduler/predicates/ha.go [hunk 2]
```diff
@@ -15,117 +15,111 @@ package predicates
...
 func (h *ha) Name() string {
 	return "HighAvailability"
 }
 
-// 1. First, we sort all the nodes we get from kube-scheduler by how many same kind of pod it contains,
-//    find the nodes that have least pods.
-// 2. When scheduling the first replicas pods, we must ensure no previous pods on the nodes.
-// 3. For later pods, we choose the nodes that have least pods.
+// 1. return the node to kube-scheduler if there is only one node and the pod's pvc is bound
+// 2. return these nodes that have least pods and its pods count is less than replicas/2 to kube-scheduler
```
Comment (onlymellb) [inline pkg/scheduler/predicates/ha.go]:
```suggestion
// 2. return these nodes that have least pods and its pods count is less than (replicas+1)/2 to kube-scheduler
```

File: pkg/scheduler/predicates/ha.go [hunk 3]
```diff
@@ -15,117 +15,111 @@ package predicates
...
 	glog.V(4).Infof("nodeMap: %+v", nodeMap)
 
-	var min int
-	var minInitialized bool
-	for _, podNameArr := range nodeMap {
-		count := len(podNameArr)
-		if !minInitialized {
-			minInitialized = true
-			min = count
+	min := -1
```
Comment (onlymellb) [inline pkg/scheduler/predicates/ha.go]: Is it better to initialize the min variable to a large enough value?

Comment (weekface) [reply pkg/scheduler/predicates/ha.go]: Large enough, how big is it?

Comment (onlymellb) [reply pkg/scheduler/predicates/ha.go]: The default number of pods that can be run on a kubelet is 110, so I think it's okay to be larger than this number, what do you think?

File: pkg/scheduler/predicates/ha.go [hunk 4]
```diff
@@ -15,117 +15,111 @@ package predicates
...
+			continue
+		}
+		if podsCount < min {
+			min = podsCount
+			minNodeNames = make([]string, 0)
 		}
+		minNodeNames = append(minNodeNames, nodeName)
 	}
 
+	if len(minNodeNames) == 0 {
```
Comment (onlymellb) [inline pkg/scheduler/predicates/ha.go]: If the passed parameter `nodes []apiv1.Node` is empty slice, should we return quickly at the beginning of the function?

File: pkg/scheduler/predicates/ha.go (hunk 3 above)
Comment (weekface) [reply pkg/scheduler/predicates/ha.go]: This number may be modified by the user. -1 is good enough i think.

File: pkg/scheduler/predicates/ha_test.go [hunk 5]
```diff
@@ -39,229 +40,327 @@ func TestMapNil(t *testing.T) {
 
 	g.Expect(m["c"] == nil).To(Equal(true))
 	g.Expect(m["a"] == nil).To(Equal(false))
+
+	var i *int
+	g.Expect(i).To(BeNil())
```
Comment (tennix) [inline pkg/scheduler/predicates/ha_test.go]: What is this?

File: pkg/scheduler/predicates/ha_test.go [hunk 6]
```diff
@@ -39,229 +40,327 @@ func TestMapNil(t *testing.T) {
...
+}
+
+func TestSortEmptyArr(t *testing.T) {
```
Comment (tennix) [inline pkg/scheduler/predicates/ha_test.go]: What's the purpose of this test?

**Pull request merged**